File: booklore-api/src/main/java/com/adityachandel/booklore/service/book/BookNoteService.java
package com.adityachandel.booklore.service.book;

import com.adityachandel.booklore.config.security.service.AuthenticationService;
import com.adityachandel.booklore.exception.ApiError;
import com.adityachandel.booklore.mapper.BookNoteMapper;
import com.adityachandel.booklore.model.dto.BookLoreUser;
import com.adityachandel.booklore.model.dto.BookNote;
import com.adityachandel.booklore.model.dto.CreateBookNoteRequest;
import com.adityachandel.booklore.model.entity.BookEntity;
import com.adityachandel.booklore.model.entity.BookLoreUserEntity;
import com.adityachandel.booklore.model.entity.BookNoteEntity;
import com.adityachandel.booklore.repository.BookNoteRepository;
import com.adityachandel.booklore.repository.BookRepository;
import com.adityachandel.booklore.repository.UserRepository;
import jakarta.persistence.EntityNotFoundException;
import lombok.AllArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.stream.Collectors;

@Service
@AllArgsConstructor
public class BookNoteService {

    private final BookNoteRepository bookNoteRepository;
    private final BookRepository bookRepository;
    private final UserRepository userRepository;
    private final BookNoteMapper mapper;
    private final AuthenticationService authenticationService;

    @Transactional(readOnly = true)
    public List<BookNote> getNotesForBook(Long bookId) {
        BookLoreUser currentUser = authenticationService.getAuthenticatedUser();
        return bookNoteRepository.findByBookIdAndUserIdOrderByUpdatedAtDesc(bookId, currentUser.getId())
                .stream()
                .map(mapper::toDto)
                .collect(Collectors.toList());
    }

    @Transactional
    public BookNote createOrUpdateNote(CreateBookNoteRequest request) {
        BookLoreUser currentUser = authenticationService.getAuthenticatedUser();

        BookEntity book = bookRepository.findById(request.getBookId())
                .orElseThrow(() -> ApiError.BOOK_NOT_FOUND.createException(request.getBookId()));

        BookLoreUserEntity user = userRepository.findById(currentUser.getId())
                .orElseThrow(() -> new EntityNotFoundException("User not found: " + currentUser.getId()));

        BookNoteEntity noteEntity;

        if (request.getId() != null) {
            noteEntity = bookNoteRepository.findByIdAndUserId(request.getId(), currentUser.getId())
                    .orElseThrow(() -> new EntityNotFoundException("Note not found: " + request.getId()));
            noteEntity.setTitle(request.getTitle());
            noteEntity.setContent(request.getContent());
        } else {
            noteEntity = BookNoteEntity.builder()
                    .user(user)
                    .book(book)
                    .title(request.getTitle())
                    .content(request.getContent())
                    .build();
        }

        BookNoteEntity savedNote = bookNoteRepository.save(noteEntity);
        return mapper.toDto(savedNote);
    }

    @Transactional
    public void deleteNote(Long noteId) {
        BookLoreUser currentUser = authenticationService.getAuthenticatedUser();
        BookNoteEntity note = bookNoteRepository.findByIdAndUserId(noteId, currentUser.getId()).orElseThrow(() -> new EntityNotFoundException("Note not found: " + noteId));
        bookNoteRepository.delete(note);
    }
}

File: booklore-ui/src/app/shared/service/book-note.service.ts
import {Injectable, inject} from '@angular/core';
import {HttpClient} from '@angular/common/http';
import {Observable} from 'rxjs';
import {API_CONFIG} from '../../core/config/api-config';

export interface BookNote {
  id: number;
  userId: number;
  bookId: number;
  title: string;
  content: string;
  createdAt: string;
  updatedAt: string;
}

export interface CreateBookNoteRequest {
  id?: number;
  bookId: number;
  title: string;
  content: string;
}

@Injectable({
  providedIn: 'root'
})
export class BookNoteService {

  private readonly url = `${API_CONFIG.BASE_URL}/api/v1/book-notes`;
  private readonly http = inject(HttpClient);

  getNotesForBook(bookId: number): Observable<BookNote[]> {
    return this.http.get<BookNote[]>(`${this.url}/book/${bookId}`);
  }

  createOrUpdateNote(request: CreateBookNoteRequest): Observable<BookNote> {
    return this.http.post<BookNote>(this.url, request);
  }

  deleteNote(noteId: number): Observable<void> {
    return this.http.delete<void>(`${this.url}/${noteId}`);
  }
}
